### PR TITLE
KOS Actor

### PR DESCRIPTION
**Pull request by sinui0**
This PR re-implements the KOS actors, removing our dependency on `xtra`. Keep in mind that these modules are already dead-on-arrival in a way, as we are going to be replacing them with SoftSpokenOT. Don't get too bogged down on all the edge-cases for now :slightly_smiling_face: 

# Changes
1. Adjusted the verifiable OT interface a bit to support parallel verification.
2. Added `Verify` as a typed state into the KOS receiver core, adding a bit of safety and saving the outer layer from having to manually distribute copies of $\Delta$
3. Implemented the actors.

### CODE REVIEW

**Comment by sinui0** [comment]
I will put tracing instrumentation into a separate PR

File: ot/mpz-ot/src/kos/sender.rs [hunk 1]
```diff
@@ -65,42 +66,9 @@ where
...
-            // Store the payload to reveal to the receiver later.
-            self.cointoss_payload = Some(payload);
-
-            seeds[0]
-        } else {
-            Block::random(&mut thread_rng())
-        };
-
-        self._setup_with_delta(sink, stream, delta).await
+    /// Returns a mutable refernce to the inner sender state.
```
Comment (th4s) [inline ot/mpz-ot/src/kos/sender.rs]: typo: `reference`

File: ot/mpz-ot/src/kos/receiver.rs [hunk 2]
```diff
@@ -65,51 +63,14 @@ where
...
-        let ext_receiver = ext_receiver.setup(seeds);
-
-        self.state = State::Extension(Box::new(ext_receiver));
+    /// Returns a reference to the inner receiver state.
+    pub(crate) fn state(&self) -> &State {
+        &self.state
+    }
 
-        Ok(())
+    /// Returns a mutable refernce to the inner receiver state.
```
Comment (th4s) [inline ot/mpz-ot/src/kos/receiver.rs]: typo: `reference`

**Review by th4s — APPROVED** [review]
Good work :rocket:

I feel it would help understanding to put the actors and the remotes (`SharedSender`, `SharedReceiver`) into different modules.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            Command::Receive {
+                id,
+                choices,
+                caller_response,
+            } => {
+                let keys = match self.start_transfer(&id, &choices).await {
+                    Ok(keys) => keys,
+                    Err(e) => {
+                        _ = caller_response.send(Err(e));
+                        return Ok(());
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:236]: is `return Ok(());` by design? Is there anything wrong with both sending an error to the caller and also returning an error from the actor?

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+                    }
+                };
+
+                self.state.ids.insert(id.clone(), keys.id());
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:247]: it's best to return an error on duplicate ids
```rust
if self.state.ids.insert(id.clone(), keys.id()).is_some() {
                    _ = caller_response.send(Err(ReceiverError::Other(format!(
                        "duplicate transfer id: {id}"
                    ))));
                    return Err(ReceiverError::Other(format!("duplicate transfer id: {id}")));
                }
```

File: ot/mpz-ot/src/kos/receiver.rs [hunk 5]
```diff
@@ -181,13 +142,126 @@ where
...
+        Si: IoSink<Message<BaseOT::Msg>> + Send + Unpin,
+        St: IoStream<Message<BaseOT::Msg>> + Send + Unpin,
+    >(
+        &mut self,
+        sink: &mut Si,
+        stream: &mut St,
+    ) -> Result<(), ReceiverError> {
+        let receiver = self.state.replace(State::Error).into_extension()?;
+
+        // Finalize cointoss to determine expected delta
```
Comment (themighty1) [inline ot/mpz-ot/src/kos/receiver.rs]: coin toss

File: ot/mpz-ot/src/kos/sender.rs [hunk 6]
```diff
@@ -226,13 +194,112 @@ where
...
+        if self.state.is_extension() {
+            return Ok(());
+        }
+
+        // Set up base OT if not already done
+        self.base
+            .setup(&mut into_base_sink(sink), &mut into_base_stream(stream))
+            .await?;
+
+        // If the sender is committed, we sample delta using a cointoss.
```
Comment (themighty1) [inline ot/mpz-ot/src/kos/sender.rs]: coin toss

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+};
+use mpz_core::{Block, ProtocolMessage};
+use mpz_ot_core::kos::{msgs::SenderPayload, PayloadRecord};
+
+use crate::actor::kos::{
+    into_kos_sink, into_kos_stream,
+    msgs::{ActorMessage, Message, TransferPayload, TransferRequest},
+    ReceiverActorError,
+};
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:27]: /// Commands which can be received by [`ReceiverActor`].

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+enum Command {
+    Receive {
+        id: String,
+        choices: Vec<bool>,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs]: /// Used to send back the `Result` to the caller of the `Receive` command.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+        caller_response: oneshot::Sender<Result<(ReceiverKeys, SenderPayload), ReceiverError>>,
+    },
+    Verify(Verify),
+    Shutdown(Shutdown),
+}
+
+struct Verify {
+    id: String,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:43]: /// Used to send back the `Result` to the caller of the `Verify` command.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    caller_response: oneshot::Sender<Result<PayloadRecord, ReceiverError>>,
+}
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:47]: /// A pending oblivious transfer which was requested by this KOS receiver but has not yet been
/// responded to by the KOS sender.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+struct PendingTransfer {
+    keys: ReceiverKeys,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:56]: /// Used to send back the `Result` to the caller of the `Receive` command.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    caller_response: oneshot::Sender<Result<(ReceiverKeys, SenderPayload), ReceiverError>>,
+}
+
+opaque_debug::implement!(PendingTransfer);
+
+struct Shutdown {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs]: /// Used to send back the `Result` to the caller of the `Shutdown` command.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+    caller_response: oneshot::Sender<Result<(), ReceiverActorError>>,
+}
+
+#[derive(Default)]
+struct State {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:64]: /// All oblivious transfer ids seen so far.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+    ids: HashMap<String, u32>,
+
+    pending_transfers: HashMap<String, PendingTransfer>,
+    pending_verify: VecDeque<Verify>,
+}
+
+/// KOS receiver actor.
+pub struct ReceiverActor<BaseOT, Si, St> {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:73]: /// A sink to send messages to the KOS sender actor.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+    sink: Si,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:75]: /// A stream to receive messages from the KOS sender actor.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+    stream: Fuse<St>,
+
+    receiver: Receiver<BaseOT>,
+
+    state: State,
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:82]: /// The sending end of the channel used by the controller to send commands to this actor.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+    command_sender: mpsc::UnboundedSender<Command>,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:84]: /// The receiving end of the channel used by this actor to receive commands from the controller.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+    pub fn receiver(&self) -> SharedReceiver {
+        SharedReceiver {
+            sender: self.command_sender.clone(),
+        }
+    }
+
+    /// Run the receiver actor.
+    pub async fn run(&mut self) -> Result<(), ReceiverActorError> {
+        loop {
+            futures::select! {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:131]: // Processes a message received from the Sender.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+                msg = self.stream.select_next_some() => self.handle_msg(msg?).await?,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:133]: // Processes a command from the controller.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+                        _ = caller_response.send(Ok(()));
+                        return Ok(());
+                    }
+
+                    self.handle_cmd(cmd).await?
+                },
+            }
+        }
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:146]: /// Starts an oblivious transfer by sending a request to the peer.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+        // Process backlog of pending verifications.
+        let backlog = std::mem::take(&mut self.state.pending_verify);
+        for verify in backlog {
+            self.handle_verify(verify)
+        }
+
+        Ok(())
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:190]: /// Handles the `Verify` command from the controller.
///
/// Sends back the [`PayloadRecord`] which the controller must verify itself.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+    fn handle_verify(&mut self, verify: Verify) {
+        // If we're ready to start verifying we do so, otherwise, we buffer
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs]: comma before "we do so" missing

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+            } else {
+                _ = caller_response.send(Err(ReceiverError::Other(format!(
+                    "transfer id not found: {id}"
+                ))));
+            }
+        } else {
+            self.state.pending_verify.push_back(verify)
+        }
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:223]: /// Handles commands received from the controller.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+                    },
+                );
+            }
+            Command::Verify(verify) => self.handle_verify(verify),
+            Command::Shutdown(_) => unreachable!("shutdown should be handled already"),
+        }
+
+        Ok(())
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:262]: /// Handles a message from the KOS sender actor.

File: ot/mpz-ot/src/actor/kos/receiver.rs [hunk 25]
```diff
@@ -0,0 +1,346 @@
...
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// KOS receiver actor control.
+#[derive(Debug, Clone)]
+pub struct SharedReceiver {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/receiver.rs:297]: /// The sending end of the channel used to send commands to the actor.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 26]
```diff
@@ -0,0 +1,330 @@
...
+    actor::kos::{
+        into_kos_sink, into_kos_stream,
+        msgs::{ActorMessage, Message, TransferPayload, TransferRequest},
+    },
+    kos::{Sender, SenderError, SenderKeys},
+    CommittedOTReceiver, CommittedOTSenderShared, OTError, OTReceiver, OTSenderShared, OTSetup,
+};
+
+use super::SenderActorError;
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:24]: /// Commands which can be received by [`SenderActor`].

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 27]
```diff
@@ -0,0 +1,330 @@
...
+enum Command {
+    GetKeys(GetKeys),
+    SendPayload(PendingPayload),
+    Shutdown(Shutdown),
+}
+
+struct GetKeys {
+    id: String,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:33]: /// Used to send back the `Result` to the caller of the `GetKeys` command.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 28]
```diff
@@ -0,0 +1,330 @@
...
+    caller_response: oneshot::Sender<Result<SenderKeys, SenderError>>,
+}
+
+struct PendingPayload {
+    id: String,
+    payload: SenderPayload,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:40]: /// Used to send back the `Result` to the caller of the `PendingPayload` command.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 29]
```diff
@@ -0,0 +1,330 @@
...
+    caller_response: oneshot::Sender<Result<(), SenderError>>,
+}
+
+struct Shutdown {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:45]: /// Used to send back the `Result` to the caller of the `Shutdown` command.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 30]
```diff
@@ -0,0 +1,330 @@
...
+#[derive(Default)]
+struct State {
+    pending_keys: HashMap<String, Result<SenderKeys, SenderError>>,
+    pending_callers: HashMap<String, oneshot::Sender<Result<SenderKeys, SenderError>>>,
+}
+
+opaque_debug::implement!(State);
+
+/// KOS sender actor.
+pub struct SenderActor<BaseOT, Si, St> {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:59]: /// A sink to send messages to the KOS receiver actor.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 31]
```diff
@@ -0,0 +1,330 @@
...
+    sink: Si,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:61]: /// A stream to receive messages from the KOS receiver actor.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 32]
```diff
@@ -0,0 +1,330 @@
...
+    stream: Fuse<St>,
+
+    sender: Sender<BaseOT>,
+
+    state: State,
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:68]: /// The sending end of the channel used by the controller to send commands to this actor.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 33]
```diff
@@ -0,0 +1,330 @@
...
+    command_sender: mpsc::UnboundedSender<Command>,
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:70]: /// The receiving end of the channel used by this actor to receive commands from the controller.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 34]
```diff
@@ -0,0 +1,330 @@
...
+
+    /// Runs the sender actor.
+    pub async fn run(&mut self) -> Result<(), SenderActorError> {
+        loop {
+            futures::select! {
+                // Processes a message received from the Receiver.
+                msg = self.stream.select_next_some() => {
+                    self.handle_msg(msg?.into_actor_message()?)?;
+                }
+                // Processes a command
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs]: // Processes a command from the controller.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 35]
```diff
@@ -0,0 +1,330 @@
...
+                        _ = caller_response.send(Ok(()));
+                        return Ok(());
+                    }
+
+                    self.handle_cmd(cmd).await;
+                }
+            }
+        }
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:149]: /// Handles commands received from the controller.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 36]
```diff
@@ -0,0 +1,330 @@
...
+    async fn handle_cmd(&mut self, cmd: Command) {
+        match cmd {
+            Command::GetKeys(GetKeys {
+                id,
+                caller_response,
+            }) => {
+                if let Some(keys) = self.state.pending_keys.remove(&id) {
+                    _ = caller_response.send(keys);
+                } else {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:159]: // The peer has not requested an OT with this id yet.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 37]
```diff
@@ -0,0 +1,330 @@
...
+                    .sink
+                    .send(ActorMessage::TransferPayload(TransferPayload { id, payload }).into())
+                    .await;
+
+                _ = caller_response.send(res.map_err(SenderError::from));
+            }
+            Command::Shutdown(_) => unreachable!("shutdown should be handled already"),
+        }
+    }
+
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:179]: /// Handles a message from the KOS receiver actor.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 38]
```diff
@@ -0,0 +1,330 @@
...
+            .reveal(
+                &mut into_kos_sink(&mut self.sink),
+                &mut into_kos_stream(&mut self.stream),
+            )
+            .await
+            .map_err(SenderActorError::from)
+    }
+}
+
+/// KOS Shared Sender
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs]: /// KOS Shared Sender controller.

File: ot/mpz-ot/src/actor/kos/sender.rs [hunk 39]
```diff
@@ -0,0 +1,330 @@
...
+#[derive(Clone)]
+pub struct SharedSender {
```
Comment (themighty1) [inline ot/mpz-ot/src/actor/kos/sender.rs:252]: /// The sending end of the channel used to send commands to the actor.

**Review by themighty1 — CHANGES_REQUESTED** [review]
gw, the first 2 review items below are questions about the code, the rest are just comments.

File: ot/mpz-ot/src/actor/kos/receiver.rs (hunk 3 above)
Comment (sinui0) [reply ot/mpz-ot/src/actor/kos/receiver.rs:236]: It's by design, I don't think the actor needs to be shut down as the error is specific to that particular call

File: ot/mpz-ot/src/actor/kos/receiver.rs (hunk 4 above)
Comment (sinui0) [reply ot/mpz-ot/src/actor/kos/receiver.rs:247]: agreed, though as above I don't think this should shut down the actor with an error

Comment (sinui0) [reply ot/mpz-ot/src/actor/kos/receiver.rs:247]: Fixed, but returns `Ok(())`

**Review by themighty1 — APPROVED** [review]

**Pull request merged**